### PR TITLE
feat(Cloud Databases): Database user password complexity validation

### PR DESCRIPTION
**Pull request by alexhemard**
From the IBM Cloud Announcement:
> **IBM Cloud Databases: User Password Complexity Policy Change**
> Please note, effective November 15, 2023, IBM Cloud Databases is changing our password complexity policy for new or changed database user passwords: Any database user passwords edited or created will be required to be a minimum of 15 characters and contain at least one letter and number.

This PR validates the updated database user password complexity during plan

- Update documentation
- Change minimum password length to 15
- Validator for `adminpassword` attribute 
- Validator in `CustomizeDiff` method for `users` attribute
- Handle `ops_manager` user updates by deleting then re-creating

Other changes:
- Adds `DatabaseUser` struct with Password validation and C**~R~**UD methods
- Refactor `userChange` struct to use `*DatabaseUser`
- Replaced `userUpdateCreate` method with separate calls to `Update()` and `Create()` 
- Renamed `check<Thing>` methods to `validate<Thing>`
- Renamed customized diff methods to `validate<Thing>Diff` 

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes https://github.com/IBM-Cloud/terraform-provider-ibm/issues/4951

Output from acceptance testing:

```
TESTARGS='-run="TestValidateUserPassword"' make test

go test -run=TestValidateUserPassword -timeout=30s -parallel=4 github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/directlink github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/dnsservices github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/enterprise 
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database	0.260s

$ make testacc TESTARGS='-run="(TestAccIBMDatabaseInstancePostgresBasic|TestAccIBMMongoDBEnterpriseDatabaseInstanceBasic)"'

...

=== RUN   TestAccIBMMongoDBEnterpriseDatabaseInstanceBasic
=== PAUSE TestAccIBMMongoDBEnterpriseDatabaseInstanceBasic
=== RUN   TestAccIBMDatabaseInstancePostgresBasic
=== PAUSE TestAccIBMDatabaseInstancePostgresBasic
=== CONT  TestAccIBMMongoDBEnterpriseDatabaseInstanceBasic
=== CONT  TestAccIBMDatabaseInstancePostgresBasic
--- PASS: TestAccIBMDatabaseInstancePostgresBasic (1150.60s)
--- PASS: TestAccIBMMongoDBEnterpriseDatabaseInstanceBasic (6419.03s)
PASS


```

### CODE REVIEW

File: ibm/service/database/resource_ibm_database.go [hunk 1]
```diff
@@ -975,50 +993,6 @@ func getGroups(instanceID string, meta interface{}) (groups []clouddatabasesv5.G
...
-	if value != resource.Allocation/nodeCount && !resource.IsAdjustable {
-		return fmt.Errorf("%s can not change %s value after create", groupId, resourceName)
-	}
-	if value < resource.Allocation/nodeCount && !resource.CanScaleDown {
-		return fmt.Errorf("can not scale %s group %s below %d to %d", groupId, resourceName, resource.Allocation/nodeCount, value)
-	}
-	return nil
-}
-
-func checkGroupValue(name string, limits GroupResource, divider int, diff *schema.ResourceDiff) error {
```
Comment (alexhemard) [inline ibm/service/database/resource_ibm_database.go:1000]: This appears unused

File: ibm/service/database/resource_ibm_database.go [hunk 2]
```diff
@@ -2401,7 +2340,10 @@ func waitForDatabaseTaskComplete(taskId string, d *schema.ResourceData, meta int
 	delayDuration := 5 * time.Second
 
 	timeout := time.After(t)
-	delay := time.Tick(delayDuration)
+	ticker := time.NewTicker(delayDuration)
+	delay := ticker.C
+	defer ticker.Stop()
```
Comment (omaraibrahim) [inline ibm/service/database/resource_ibm_database.go:2345]: Why are we changing the delay logic?

File: ibm/service/database/resource_ibm_database.go [hunk 3]
```diff
@@ -2748,7 +2690,28 @@ func expandGroups(_groups []interface{}) []*Group {
...
+	}
+	if resource.StepSize == 0 {
+		return fmt.Errorf("%s group must have members scaled > 0 before scaling %s", groupId, resourceName)
+	}
+	if value < resource.Minimum/nodeCount || value > resource.Maximum/nodeCount || value%(resource.StepSize/nodeCount) != 0 {
+		if !(value == 0 && resource.IsOptional) {
+			return fmt.Errorf("%s group %s must be >= %d and <= %d in increments of %d", groupId, resourceName, resource.Minimum/nodeCount, resource.Maximum/nodeCount, resource.StepSize/nodeCount)
+		}
+	}
+	if value != resource.Allocation/nodeCount && !resource.IsAdjustable {
```
Comment (omaraibrahim) [inline ibm/service/database/resource_ibm_database.go:2705]: We may want to add comments on top of each of these if statements. It's a little difficult to see why these are needed without having a strong understanding of the way we manage scaling.

File: ibm/service/database/resource_ibm_database.go (hunk 2 above)
Comment (alexhemard) [reply ibm/service/database/resource_ibm_database.go:2345]: This got left in while debugging `waitForDatabaseTaskComplete` but allows the Ticker to be garbage collected see https://pkg.go.dev/time#Tick

**Review by omaraibrahim — APPROVED** [review]
LGTM

File: ibm/service/database/resource_ibm_database.go [hunk 4]
```diff
@@ -2836,73 +2799,264 @@ func checkV5Groups(_ context.Context, diff *schema.ResourceDiff, meta interface{
...
+	return c.Old != nil && c.New == nil
+}
+
+func (c *userChange) isCreate() bool {
+	return c.Old == nil && c.New != nil
+}
+
+func (c *userChange) isUpdate() bool {
+	return c.New != nil && c.Old != nil && c.Old.Password != c.New.Password
+}
```
Comment (obai-1) [inline ibm/service/database/resource_ibm_database.go:2882]: This considers password change as an update. What about changes to name/type/role ?

I understand some of them will cause delete and recreate. How are they handled?

Comment (alexhemard) [reply ibm/service/database/resource_ibm_database.go:2882]: Users are grouped by name and type in the `expandUserChanges` method. Any change to name or type will delete the old user and create a new one. 

Ops Manager role updates are not currently supported in the provider but updating roles will be added with Redis RBAC support.

**Comment by ocofaigh** [comment]
I don't see this in the 1.60.0 release notes, so I'm guessing it will be part of 1.61.0 ?

**Pull request merged**